File: HW7/Hom4/Form1.vb
Public Class Form1

    Private Bitmap As Bitmap
    Private Graphics As Graphics
    Private Bitmap2 As Bitmap
    Private Graphics2 As Graphics
    Private rand As Random = New Random()
    Private PenAbso As Pen = New Pen(Color.LightGreen, 2)
    Dim interarrivals As List(Of Integer)
    Dim interarrivals_distribution_dictionary As SortedDictionary(Of Integer, Integer)


    Private Function tranX(ByVal x As Double, ByVal minX As Double, ByVal maxX As Double, ByVal w As Integer) As Integer
        Return CInt((w * (x - minX) / (maxX - minX)))
    End Function

    Private Function tranY(ByVal y As Double, ByVal minY As Double, ByVal maxY As Double, ByVal h As Integer) As Integer
        Return CInt((h - h * (y - minY) / (maxY - minY)))
    End Function


    Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click

        Bitmap = New Bitmap(PictureBox1.Width, PictureBox1.Height)
        Graphics = Graphics.FromImage(Bitmap)
        Graphics.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality
        Graphics.Clear(Color.Black)
        Dim trials = TrackBar1.Value
        Dim lambda = TrackBar3.Value
        interarrivals = New List(Of Integer)
        Dim count_interarrivals = 0
        Dim rectangle = New Rectangle(20, 20, Bitmap.Width, Bitmap.Height)
        'Graphics.FillRectangle(Brushes.Black, rectangle)
        'Graphics.DrawRectangle(Pens.Black, rectangle)
        RichTextBox1.Clear()

        Dim probability As Double = (lambda / trials) * 100
        RichTextBox1.AppendText("probability: " & probability)
        RichTextBox1.AppendText(Environment.NewLine)


        Dim minX_a As Double = 0
        Dim maxX_a As Double = CDbl(trials)
        Dim minY_a As Double = 0
        Dim maxY_a As Double = CDbl(trials)


        Dim y_a As Double = 0
        Dim points_a As List(Of Point) = New List(Of Point)
        For tria_num As Integer = 0 To trials - 1
            Dim random_number = rand.Next(0, 100)
            If random_number <= probability Then
                y_a = y_a + 1
                interarrivals.Add(count_interarrivals)
                count_interarrivals = 0
            Else
                count_interarrivals = count_interarrivals + 1
            End If
            'Absolute
            Dim cord_x_a As Integer = tranX(tria_num, minX_a, maxX_a, rectangle.Width)
            Dim cord_y_a As Integer = tranY(y_a, minY_a, maxY_a, rectangle.Height)
            Dim p_a As Point = New Point(cord_x_a, cord_y_a)
            points_a.Add(p_a)
            'RichTextBox1.AppendText(tria_num.ToString & ": " & y_a)
            'RichTextBox1.AppendText(Environment.NewLine)
        Next
        Graphics.DrawLines(PenAbso, points_a.ToArray)
        PictureBox1.Image = Bitmap
        interarrivals_distribution()
        draw_interarrivals_distribuiton()
    End Sub

    Private Sub interarrivals_distribution()
        interarrivals_distribution_dictionary = New SortedDictionary(Of Integer, Integer)
        For Each elem In interarrivals
            'RichTextBox1.AppendText("elem: " & elem)
            'RichTextBox1.AppendText(Environment.NewLine)
            If Not interarrivals_distribution_dictionary.ContainsKey(elem) Then
                interarrivals_distribution_dictionary.Add(elem, 1)

            Else
                interarrivals_distribution_dictionary(elem) += 1
            End If
            'RichTextBox1.AppendText("elem_val: " & interarrivals_distribution_dictionary(elem))
            'RichTextBox1.AppendText(Environment.NewLine)
        Next
    End Sub

    Private Sub draw_interarrivals_distribuiton()
        Bitmap2 = New Bitmap(PictureBox2.Width, PictureBox2.Height)
        Graphics2 = Graphics.FromImage(Bitmap2)
        Graphics2.Clear(Color.Black)
        Dim rectangle2 = New Rectangle(0, 0, Bitmap2.Width, Bitmap2.Height)

        Dim max_value = interarrivals_distribution_dictionary.Values.Max
        Dim space_height = rectangle2.Bottom - rectangle2.Top
        Dim space_width = rectangle2.Right - rectangle2.Left
        Dim number_of_keys = interarrivals_distribution_dictionary.Keys.Count
        Dim bar_width = space_width / number_of_keys

        Dim start = rectangle2.X

        For Each key In interarrivals_distribution_dictionary
            Dim rect_height = key.Value / max_value * space_height
            Dim bar_rect = New Rectangle(start, rectangle2.Bottom - rect_height, bar_width, rect_height)

            Graphics2.FillRectangle(Brushes.Green, bar_rect)
            Graphics2.DrawRectangle(Pens.White, bar_rect)
            start += bar_width

            RichTextBox1.AppendText("key: " & key.Key)
            RichTextBox1.AppendText(Environment.NewLine)
            RichTextBox1.AppendText("value: " & key.Value)
            RichTextBox1.AppendText(Environment.NewLine)
        Next




        PictureBox2.Image = Bitmap2
    End Sub

    Private Sub TrackBar1_Scroll(sender As Object, e As EventArgs) Handles TrackBar1.Scroll
        Label1.Text = "Trials: " & TrackBar1.Value.ToString
    End Sub


    Private Sub TrackBar3_Scroll(sender As Object, e As EventArgs) Handles TrackBar3.Scroll
        Label3.Text = "Lambda: " & TrackBar3.Value.ToString
    End Sub

    Private Sub Label1_Click(sender As Object, e As EventArgs) Handles Label1.Click

    End Sub
End Class

File: HW4/Homework4/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using static System.Windows.Forms.VisualStyles.VisualStyleElement;
namespace Homework4
{
    
    public partial class Form1 : Form
    {
        internal class Interval
        {
            public int up;
            public int down;

            public Interval(int down, int up)
            {
                this.up = up;
                this.down = down;
            }

            public override string ToString()
            {
                return down.ToString() + " - " + up.ToString();
            }
        }
        Bitmap b;
        Graphics g;
        Random random = new Random();
        Pen PenTrajectory = new Pen(Color.OrangeRed, 1);
        double successProbability = 0;

        public Form1()
        {
            InitializeComponent();
        }

        private void button1_Click(object sender, EventArgs e)
        {
            successProbability = (double)numericUpDown3.Value;
            b = new Bitmap(pictureBox1.Width, pictureBox1.Height);
            g = Graphics.FromImage(b);
            g.SmoothingMode = System.Drawing.Drawing2D.SmoothingMode.HighQuality;
            g.Clear(Color.White);

            int TrialsCount = (int)numericUpDown1.Value;
            int TrajectoryNumber = (int)numericUpDown2.Value;

            double minX = 0;
            double maxX = (double)TrialsCount;
            double minY = 0;
            double maxY = 0;
            if (checkBox3.Checked == true) maxY = (double)TrialsCount;
            if (checkBox1.Checked == true) maxY = 1;
            if (checkBox2.Checked == true) maxY = ((double)TrialsCount) / Math.Sqrt((double)TrialsCount);
            double Y = 0;

            Rectangle r = new Rectangle(20, 20, b.Width - 300, b.Height - 40);
            g.FillRectangle(Brushes.Black, r);
            g.DrawRectangle(Pens.Black, r);
            List<Point> lastpoints = new List<Point>();
            for (int t = 0; t < TrajectoryNumber; t++)
            {
                List<Point> points = new List<Point>();
                double Y1 = 0;
                for (int X = 0; X < TrialsCount; X++)
                {
                    random.NextDouble();

                    if (random.NextDouble() < successProbability) Y1 = Y1 + 1;
                    if (checkBox3.Checked == true) Y = Y1;
                    if (checkBox1.Checked == true) Y = Y1 / (X + 1);
                    if (checkBox2.Checked == true) Y = Y1 / Math.Sqrt(X + 1);
                    int xCord = linearTransformX(X, minX, maxX, r.Left, r.Width);
                    int yCord = linearTransformY(Y, minY, maxY, r.Top, r.Height);

                    Point p = new Point(xCord, yCord);
                    points.Add(p);

                    if (X == TrialsCount - 1)
                    {
                        lastpoints.Add(p);
                    }
                }
                g.DrawLines(PenTrajectory, points.ToArray());
            }
            int min_y = lastpoints.Min(p => p.Y);
            int max_y = lastpoints.Max(p => p.Y);

            Rectangle r2 = new Rectangle(r.Right + 10, 20, 260, b.Height - 40);
            g.FillRectangle(Brushes.Black, r2);
            g.DrawRectangle(Pens.Black, r2);
            if (TrajectoryNumber == 1)
            {
                foreach (Point p in lastpoints)
                {
                    Rectangle re = new Rectangle(r2.Left + 10, p.Y - 10, r2.Right - r2.Left - 20, 20);
                    g.FillRectangle(Brushes.Orange, re);
                    g.DrawRectangle(Pens.White, re);
                }
            }
            else
            {
                int intervals_number = TrajectoryNumber / 2;
                if (intervals_number > 15)
                {
                    intervals_number = 15;
                }
                else if (intervals_number <= 0)
                {
                    intervals_number = 1;
                }

                int span = max_y - min_y;
                int interval_size = (max_y - min_y) / intervals_number;

                while (min_y + interval_size * intervals_number < max_y + 1)
                {
                    intervals_number++;
                }

                int minimo = min_y;

                Dictionary<Interval, int> intervalli = new Dictionary<Interval, int>();

                for (int j = 0; j < intervals_number; j++)
                {
                    Interval intervallo = new Interval(minimo, minimo + interval_size);
                    minimo = minimo + interval_size;
                    intervalli[intervallo] = 0;
                }
                foreach (Point p in lastpoints)
                {
                    List<Interval> inter = intervalli.Keys.ToList();
                    int intY = (int)p.Y;

                    foreach (Interval i in inter)
                    {
                        if (intY >= i.down && intY < i.up)
                        {
                            intervalli[i]++;
                        }
                    }
                }
                List<Rectangle> chart = new List<Rectangle>();

                int dimdisp = r2.Right - r2.Left - 20; 
                int maxValue = intervalli.Values.Max();

                foreach (var v in intervalli)
                {
                    double intensity = ((double)v.Value / (double)maxValue) * dimdisp;
                    Rectangle rect = new Rectangle(r2.Left + 10, v.Key.down, (int)intensity, interval_size);
                    chart.Add(rect);
                }

                foreach (Rectangle re in chart)
                {
                    g.FillRectangle(Brushes.Orange, re);
                    g.DrawRectangle(Pens.White, re);
                }
                pictureBox1.Image = b;
            }
            pictureBox1.Image = b;
        }
            public int linearTransformX(double X, double minX, double maxX, int Left, int W)
            {
                return Left + (int)(W * ((X - minX) / (maxX - minX)));
            }

            public int linearTransformY(double Y, double minY, double maxY, int Top, int H)
            {
                return Top + (int)(H - H * ((Y - minY) / (maxY - minY)));
            }

        private void checkBox2_Click(object sender, EventArgs e)
        {
            checkBox3.Checked = false;
            checkBox1.Checked = false;
        }

        private void checkBox3_Click(object sender, EventArgs e)
        {
            checkBox1.Checked = false;
            checkBox2.Checked = false;
        }

        private void checkBox1_Click(object sender, EventArgs e)
        {
            checkBox3.Checked = false;
            checkBox2.Checked = false;
        }
    }
    
     
}
